### ISSUE TITLE
Very low benchmarks results compared to results on README.md

### ISSUE THREAD

**Issue by darxis** [chronoxor/NetCoreServer#48]
I have run the benchmarks on my PC and experiencing very low results compared too the ones listed on the README.md. I am running the 1777931c commit
Environment:
```
CPU architecutre: Intel(R) Core(TM) i7-6820HK CPU @ 2.70GHz
CPU logical cores: 8
CPU physical cores: 4
CPU clock speed: 2.70 GHz
CPU Hyper-Threading: enabled
RAM total: 16 GB
RAM free: +-6 GB

OS version: Microsoft Windows 10 Pro (build 1909), 64-bit
OS bits: 64-bit
Process bits: 64-bit
Process configuration: Release
```

Here are my results for the SSL Multicast -c 100 test:
```
Server address: 127.0.0.1
Server port: 53698
Working clients: 100
Message size: 32
Seconds to benchmarking: 10

Clients connecting...Done!
All clients connected!
Benchmarking...Done!
Clients disconnecting...Done!
All clients disconnected!

Errors: 0

Total time: 11.389 s
Total data: 5.233 MiB <---------------------------
Total messages: 171306
Data throughput: 470.038 KiB/s <---------------------------
Message latency: 66.484 mcs <---------------------------
Message throughput: 15041 msg/s <---------------------------
Press any key to exit...
```

Compared to the ones listed on the README.md:
```
Server address: 127.0.0.1
Server port: 2222
Working clients: 100
Message size: 32

Errors: 0

Total time: 10.278 s
Total data: 66.540 MiB <---------------------------
Total messages: 2179997
Data throughput: 6.483 MiB/s <---------------------------
Message latency: 4.715 mcs <---------------------------
Message throughput: 212083 msg/s <---------------------------
```

As you can see, my tests benchmark results are more than 90% lower.

Could you please re-run the benchmarks again and post your results so that I can see if there is a problem with my environment or not? Commit 1777931c

**Comment by darxis** [comment]
My iperf3 benchmark results are showing 1Gbits+/sec:
```
C:\Users\XXX\Downloads\iperf-3.1.3-win64\iperf-3.1.3-win64>iperf3 -c 127.0.0.1 -t 100 -i 5
Connecting to host 127.0.0.1, port 5201
[  4] local 127.0.0.1 port 1036 connected to 127.0.0.1 port 5201
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-5.01   sec   629 MBytes  1.05 Gbits/sec
[  4]   5.01-10.01  sec   631 MBytes  1.06 Gbits/sec
[  4]  10.01-15.00  sec   540 MBytes   906 Mbits/sec
[  4]  15.00-20.00  sec   653 MBytes  1.10 Gbits/sec
[  4]  20.00-25.00  sec   643 MBytes  1.08 Gbits/sec
[  4]  25.00-30.01  sec   436 MBytes   730 Mbits/sec
[  4]  30.01-35.00  sec   390 MBytes   655 Mbits/sec
[  4]  35.00-40.00  sec  1.01 GBytes  1.74 Gbits/sec
[  4]  40.00-45.00  sec  2.34 GBytes  4.02 Gbits/sec
[  4]  45.00-50.00  sec  4.02 GBytes  6.92 Gbits/sec
[  4]  50.00-55.00  sec  4.16 GBytes  7.15 Gbits/sec
[  4]  55.00-60.00  sec  4.33 GBytes  7.43 Gbits/sec
[  4]  60.00-65.00  sec  3.91 GBytes  6.71 Gbits/sec
[  4]  65.00-70.00  sec  3.45 GBytes  5.93 Gbits/sec
[  4]  70.00-75.00  sec  3.06 GBytes  5.25 Gbits/sec
[  4]  75.00-80.00  sec  2.82 GBytes  4.84 Gbits/sec
[  4]  80.00-85.00  sec  4.38 GBytes  7.53 Gbits/sec
[  4]  85.00-90.00  sec  3.70 GBytes  6.36 Gbits/sec
[  4]  90.00-95.00  sec  4.02 GBytes  6.90 Gbits/sec
[  4]  95.00-100.00 sec  3.94 GBytes  6.76 Gbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-100.00 sec  49.0 GBytes  4.21 Gbits/sec                  sender
[  4]   0.00-100.00 sec  49.0 GBytes  4.21 Gbits/sec                  receiver

iperf Done.
```

**Comment by chronoxor** [comment]
Commit 1777931 did not break the performance itself. But I've just run benchmarks on my environment and get the following results for `SslMulticastClient.exe -c 100 -t 4`:
```
Server address: 127.0.0.1
Server port: 2222
Working clients: 100
Message size: 32
Seconds to benchmarking: 10

Errors: 0

Total time: 10.383 s
Total data: 19.706 MiB
Total messages: 645190
Data throughput: 1.917 MiB/s
Message latency: 16.093 mcs
Message throughput: 62137 msg/s
```

I'll try to profile the benchmark and investigate why the performance slows down.

**Comment by chronoxor** [comment]
Profiler shows the hot path to `System.Net.Security.SslStream::BeginWrite()` call. As designed, so currently I don't see any paths to optimize. I think I'll rebuild the lib in Release mode and re-run all benchmarks and update the results in README.md to the current actual state!

![image](https://user-images.githubusercontent.com/4397805/76764103-5a45a580-67a5-11ea-8e71-a43296f4665f.png)

**Comment by darxis** [comment]
I think we should check at what commit the initial benchmarks were run and see what changed that caused such a performance drop.

**Comment by chronoxor** [comment]
The problem that initial version of the NetCoreServer was based on .NET Core 2.0. However I think it's not a big deal to downgrade it locally.

**Comment by darxis** [comment]
But how downgrading to .NET Core 2.0 would increase the performance 3x times?

**Comment by chronoxor** [comment]
Just build and test NetCoreServer from 30 Jan 2019 on my current environment, which results went to README.md. And get almost the same results as above:
```
CPU architecutre: Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
CPU logical cores: 8
CPU physical cores: 4
CPU clock speed: 3.998 GHz
CPU Hyper-Threading: enabled
RAM total: 31.962 GiB
RAM free: 24.129 GiB

Server address: 127.0.0.1
Server port: 2222
Working clients: 100
Message size: 32

Errors: 0

Multicast time: 10.370 s
Total data: 18.654 MiB
Total messages: 610776
Data throughput: 1.816 MiB/s
Message latency: 16.979 mcs
Message throughput: 58895 msg/s
```

**Comment by chronoxor** [comment]
I'll re-run all benchmarks and update README.md with actual results, which you could compare on your side as well.

**Comment by darxis** [comment]
Thanks @chronoxor
My story is that I am developing my custom SSL TCP client/server and I can't  send more than 10-15k messages per second, so I have found NetCoreServer stating that it can achieve much better performance but when running the benchmark I can't still achieve more than 15k messages per second. I have been tuning my code performance and profiling for 2 weeks now (including Garbage Collection delays), so it could be an issue with the environment.

Do you have an idea why your benchmark results could be 4x faster than mine?

**Comment by chronoxor** [comment]
Do you test on a localhost environment or real LAN network?

**Comment by darxis** [comment]
I am running the client and server on 127.0.0.1

**Comment by erictuvesson** [comment]
I believe that .NET Core have issues with local DNS lookup, I think I have read about it on some github issue over there.

**Comment by darxis** [comment]
@erictuvesson Thank sfor the suggestion. I have already inspected the network traffic with Wireshark and it shows that it passes through 127.0.0.1

**Comment by darxis** [comment]
Well, this is getting very strange. I have run the benchmark on two different PCs connected through 1Gbps LAN and I am getting 10x better results than running on 127.0.0.1. Please see the results:
```
SslMulticastClient.exe -a 192.168.0.10
Server address: 192.168.0.10
Server port: 53698
Working clients: 100
Message size: 32
Seconds to benchmarking: 10

Clients connecting...Done!
All clients connected!
Benchmarking...Done!
Clients disconnecting...Done!
All clients disconnected!

Errors: 0

Total time: 12.378 s
Total data: 43.269 MiB <---------------------------
Total messages: 1417643
Data throughput: 3.506 MiB/s <---------------------------
Message latency: 8.731 mcs <---------------------------
Message throughput: 114523 msg/s <---------------------------
```
As you can see, on LAN I am getting 10x better results than on 127.0.0.1.

**Comment by chronoxor** [comment]
Another idea is to perform test using IP address of your PC. Allowing packets go through your network card driver instead of windows loopback interface. 

![image](https://user-images.githubusercontent.com/4397805/76808714-a3274980-67f9-11ea-8607-c27d7f4badf3.png)

**Comment by chronoxor** [comment]
Also I've just updated benchmarks results in README.md form my home PC. Results are worse than prev taken from much more powerful environment which is not available anymore.

**Comment by darxis** [comment]
@chronoxor I have just done this, the results are still poor when I enter my LAN IP... But it could be because of Windows detecting that the PC owns this IP address and still send packets through the loopback interface

**Comment by darxis** [comment]
@chronoxor Thanks! Could you also please update the environment in the README.md?

**Comment by chronoxor** [comment]
Please note that Client app consume much more CPU and network resources than server. So it it also not perfect to measure server performance on a single host. If you run server on a LAN and connect some Client apps from different hosts you'll see that Server could handle many connections without troubles and CPU burnout.

**Comment by darxis** [comment]
I don't think the CPU is the bottleneck as Client is using 4% CPU, Server 12% CPU

**Comment by darxis** [comment]
> 
> 
> @chronoxor I have just done this, the results are still poor when I enter my LAN IP... But it could be because of Windows detecting that the PC owns this IP address and still send packets through the loopback interface

I have just checked this, it goes through the loopback adapter as in wireshark I can see the packets in the loopback adapter, not the Ethernet one

**Comment by erictuvesson** [comment]
Might be related: https://github.com/dotnet/runtime/issues/31085

**Comment by darxis** [comment]
@erictuvesson Thanks. This issue is about slow connection times when you specify localhost instead of 127.0.0.1, not my case :/

**Comment by darxis** [comment]
I haven't managed to pass the traffic through the physical NIC...

However I have run two other benchmark tools: iperf3 and CppServer cppserver-performance-ssl_multicast_client.exe test, here are the results:

Iperf3 with 32 message size (to reflect SSlMulticast benchmarks):
```
C:\Users\XXX\Downloads\iperf-3.1.3-win64>iperf3 -c 127.0.0.1 -l 32
Connecting to host 127.0.0.1, port 5201
[  4] local 127.0.0.1 port 23977 connected to 127.0.0.1 port 5201
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-1.00   sec   558 KBytes  4.57 Mbits/sec
[  4]   1.00-2.00   sec   688 KBytes  5.64 Mbits/sec
[  4]   2.00-3.00   sec   612 KBytes  5.01 Mbits/sec
[  4]   3.00-4.00   sec   358 KBytes  2.93 Mbits/sec
[  4]   4.00-5.00   sec   330 KBytes  2.70 Mbits/sec
[  4]   5.00-6.00   sec   416 KBytes  3.41 Mbits/sec
[  4]   6.00-7.00   sec   416 KBytes  3.41 Mbits/sec
[  4]   7.00-8.00   sec   352 KBytes  2.88 Mbits/sec
[  4]   8.00-9.00   sec   416 KBytes  3.41 Mbits/sec
[  4]   9.00-10.00  sec   416 KBytes  3.41 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-10.00  sec  4.45 MBytes  3.74 Mbits/sec                  sender
[  4]   0.00-10.00  sec  3.78 MBytes  3.17 Mbits/sec                  receiver

iperf Done.
```

CppServer SslMulticastClient
```
C:\Users\XXX\Source\Repos\CppServer\bin>cppserver-performance-ssl_multicast_client.exe -c 100 -p 53982
Server address: 127.0.0.1
Server port: 53982
Working threads: 4
Working clients: 100
Message size: 32
Seconds to benchmarking: 10

Asio service starting...Done!
Clients connecting...Done!
All clients connected!
Benchmarking...Done!
Clients disconnecting...Done!
All clients disconnected!
Asio service stopping...Done!

Errors: 0

Total time: 10.702 s
Total data: 975.532 MiB
Total messages: 31965842
Data throughput: 91.151 MiB/s
Message latency: 334 ns
Message throughput: 2986732 msg/s
```

As you can see iperf3 throughput is approx 3.17 Mbps and CppServer SslMulticastClient is 91MB/s = 728Mbps. Its strange that iperf3 is such slower than CppServer SslMulticastClient.

I think I could have a configuration problem that triggers in NetCoreServer SslMulticastClient and iperf3 but not triggers on CppServer SslMulticastClient

**Comment by chronoxor** [comment]
You can also try to experiment with combinations of C++ Server, Client and .NET Server, Client. They are very similar designed and compatible. Perhaps you can get more info and stats.

**Comment by chronoxor** [comment]
My C++ results are:
```
Z:\projects\Projects\CppServer\bin>cppserver-performance-ssl_multicast_client.exe -c 100
Server address: 127.0.0.1
Server port: 2222
Working threads: 4
Working clients: 100
Message size: 32
Seconds to benchmarking: 10

Asio service starting...Done!
Clients connecting...Done!
All clients connected!
Benchmarking...Done!
Clients disconnecting...Done!
All clients disconnected!
Asio service stopping...Done!

Errors: 0

Total time: 11.305 s
Total data: 3.051 GiB
Total messages: 102358287
Data throughput: 276.306 MiB/s
Message latency: 110 ns
Message throughput: 9053763 msg/s
```

**Comment by darxis** [comment]
What CPU and RAM specs do you have?

**Comment by chronoxor** [comment]
The same as in README.md:
```
===============================================================================
CPU architecutre: Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
CPU logical cores: 8
CPU physical cores: 4
CPU clock speed: 3.998 GHz
CPU Hyper-Threading: enabled
RAM total: 31.962 GiB
RAM free: 24.453 GiB
===============================================================================
OS version: Microsoft Windows 8 Enterprise Edition (build 9200), 64-bit
OS bits: 64-bit
Process bits: 64-bit
Process configuaraion: release
Local timestamp: Tue Mar 17 14:45:36 2020
UTC timestamp: Tue Mar 17 11:45:36 2020
===============================================================================
```

**Comment by darxis** [comment]
Ok I think I have found somethink. I analysed the two benchmarks in Wireshark and this is what I have found:
iperf3:
![image](https://user-images.githubusercontent.com/10614436/76853307-4eadb980-684d-11ea-8a83-3a05a4eb863f.png)

CppServer SslMulticastClient
![image](https://user-images.githubusercontent.com/10614436/76853420-83217580-684d-11ea-9c9d-44cfef32c8ff.png)

As you can see, iperf3 wait for the ACK before sending the next packets, CppServer does not wait

**Comment by chronoxor** [comment]
That's because CppServer and NetCoreServer designed to be async. Thats's how 'SslServer.Multicast()' method implemented:
```
        /// <summary>
        /// Multicast data to all connected clients
        /// </summary>
        /// <param name="buffer">Buffer to multicast</param>
        /// <param name="offset">Buffer offset</param>
        /// <param name="size">Buffer size</param>
        /// <returns>'true' if the data was successfully multicasted, 'false' if the data was not multicasted</returns>
        public virtual bool Multicast(byte[] buffer, long offset, long size)
        {
            if (!IsStarted)
                return false;

            if (size == 0)
                return true;

            // Multicast data to all sessions
            foreach (var session in Sessions.Values)
                session.SendAsync(buffer, offset, size);

            return true;
        }
``

**Comment by darxis** [comment]
You're right, that must be it.

Coming back to your benchmark results:
NetCoreServer:
```

Multicast time: 10.370 s
Total data: 18.654 MiB
Total messages: 610776
Data throughput: 1.816 MiB/s
Message latency: 16.979 mcs
Message throughput: 58895 msg/s
```

CppServer:
```

Total time: 11.305 s
Total data: 3.051 GiB
Total messages: 102358287
Data throughput: 276.306 MiB/s
Message latency: 110 ns
Message throughput: 9053763 msg/s
```
As you can see the CppServer is more than 100x faster. Isn't it a little big difference? Before updating your README.md you had much closer difference, so maybe some commit have significally slowed down the NetCoreServer?

**Comment by darxis** [comment]
> 
> 
> That's because CppServer and NetCoreServer designed to be async. Thats's how `SslServer.Multicast()` method implemented:
> 
> ```
>         /// <summary>
>         /// Multicast data to all connected clients
>         /// </summary>
>         /// <param name="buffer">Buffer to multicast</param>
>         /// <param name="offset">Buffer offset</param>
>         /// <param name="size">Buffer size</param>
>         /// <returns>'true' if the data was successfully multicasted, 'false' if the data was not multicasted</returns>
>         public virtual bool Multicast(byte[] buffer, long offset, long size)
>         {
>             if (!IsStarted)
>                 return false;
> 
>             if (size == 0)
>                 return true;
> 
>             // Multicast data to all sessions
>             foreach (var session in Sessions.Values)
>                 session.SendAsync(buffer, offset, size);
> 
>             return true;
>         }
> ```

I can't agree now, this is the wireshark traffic for NetCoreServer:
![image](https://user-images.githubusercontent.com/10614436/76854916-45721c00-6850-11ea-8e24-a9b01a4c14c1.png)

The previous traffic was for CppServer

**Comment by darxis** [comment]
Results for cross-benchmarking NetCoreServer and CppServer:
NetCoreServer Client <-> CppServer Server
```
Total time: 13.497 s
Total data: 1.050 GiB
Total messages: 35207326
Data throughput: 79.618 MiB/s
Message latency: 383 ns
Message throughput: 2608456 msg/s
```
CppServer Client <-> NetCoreServer Server
```
Total time: 10.716 s
Total data: 4.145 MiB
Total messages: 135722
Data throughput: 395.784 KiB/s
Message latency: 78.960 mcs
Message throughput: 12664 msg/s
```

We can see that the NetCoreServer SslMulticastServer is the one which has poor performance when compared to CppServer, SslMulticastClient is OK.

**Comment by darxis** [comment]
Could be a related issue: robinrodricks/FluentFTP#354

**Comment by darxis** [comment]
Solved with #49
